pkg/networking/session.go

```diff
@@ -0,0 +1,429 @@
+package networking
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/wavesplatform/gowaves/pkg/execution"
+)
+
+// Session is used to wrap a reliable ordered connection.
+type Session struct {
+	g      *execution.TaskGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config *Config
+	logger *slog.Logger
+	tp     *timerPool
+
+	conn    io.ReadWriteCloser // conn is the underlying connection
+	bufRead *bufio.Reader      // buffered reader wrapped around the connection
+
+	receiveLock   sync.Mutex    // Guards the receiveBuffer.
+	receiveBuffer *bytes.Buffer // receiveBuffer is used to store the incoming data.
+
+	sendLock sync.Mutex       // Guards the sendCh.
+	sendCh   chan *sendPacket // sendCh is used to send data to the connection.
+
+	receiving   atomic.Bool // Indicates that receiveLoop already running.
+	established atomic.Bool // Indicates that incoming Handshake was successfully accepted.
+	shutdown    sync.Once   // shutdown is used to safely close the Session.
+}
+
+// NewSession is used to construct a new session.
+func newSession(ctx context.Context, config *Config, conn io.ReadWriteCloser, tp *timerPool) (*Session, error) {
+	if config.protocol == nil {
+		return nil, ErrInvalidConfigurationNoProtocol
+	}
+	if config.handler == nil {
+		return nil, ErrInvalidConfigurationNoHandler
+	}
+	if config.keepAlive && config.keepAliveInterval <= 0 {
+		return nil, ErrInvalidConfigurationNoKeepAliveInterval
+	}
+	if config.connectionWriteTimeout <= 0 {
+		return nil, ErrInvalidConfigurationNoWriteTimeout
+	}
+	if tp == nil {
+		return nil, ErrEmptyTimerPool
+	}
+
+	sCtx, cancel := context.WithCancel(ctx)
+	s := &Session{
+		g:       execution.NewTaskGroup(suppressContextCancellationError),
+		ctx:     sCtx,
+		cancel:  cancel,
+		config:  config,
+		tp:      tp,
+		conn:    conn,
+		bufRead: bufio.NewReader(conn),
+		sendCh:  make(chan *sendPacket, 1), // TODO: Make the size of send channel configurable.
+	}
+
+	if config.slogHandler == nil {
+		config.slogHandler = discardingHandler{}
+	}
+
+	sa := [...]any{
+		slog.String("namespace", Namespace),
+		slog.String("remote", s.RemoteAddr().String()),
+	}
+	attrs := append(sa[:], config.attributes...)
+	s.logger = slog.New(config.slogHandler).With(attrs...)
```

Comment:
I think it's better to create a variable for `slogHandler` than modifying the `config`
```suggestion
	slogHandler := config.slogHandler
	if slogHandler == nil {
		slogHandler = discardingHandler{}
	}

	sa := [...]any{
		slog.String("namespace", Namespace),
		slog.String("remote", s.RemoteAddr().String()),
	}
	attrs := append(sa[:], config.attributes...)
	s.logger = slog.New(slogHandler).With(attrs...)
```